### ISSUE TITLE
File stat() modtime precision doesn't seem to include microseconds/nanoseconds

### ISSUE THREAD

**Issue by tomekit** [dart-lang/sdk#56626]
I am calling `statSync()` on file and getting modtime with ms precision: `2024-09-02 09:57:59.602`.

If I call bash CLI: `stat <filename>` I get full ns precision: `2024-09-02 09:57:59.602813345`. I am using EXT4 filesystem.

By any chance is it possible to get the modtime with full available precision down to the nanosecond?

Dart/Flutter:
![Screenshot from 2024-09-02 10-03-02](https://github.com/user-attachments/assets/27d3caf0-9362-49e7-a1e0-3e8a3ba2fab7)

Bash `stat`:
![Screenshot from 2024-09-02 10-03-29](https://github.com/user-attachments/assets/5386d318-8b9a-4717-ae80-294fe57e55a4)

* Dart version and tooling diagnostic info (`dart info`)
```
- Dart 3.4.4 (stable) (Wed Jun 12 15:54:31 2024 +0000) on "linux_x64"
- on linux / Linux 6.8.0-41-generic #41-Ubuntu SMP PREEMPT_DYNAMIC Fri Aug  2 20:41:06 UTC 2024
- locale is en_US.UTF-8
```

* Whether you are using Windows, macOS, or Linux (if applicable)
```
Ubuntu 24.04 LTS
```

**Comment by a-siva** [comment]
@tomekit supporting precision down to the nanosecond would mean we would have to change Dart's DateTime class (https://api.flutter.dev/flutter/dart-core/DateTime-class.html) to include a DateTime.fromNanosecondsSinceEpoch class. This is a breaking change. A much easier change would be for us to support microseconds precision from the statSync call, this would be supported on platforms that support microsecond precision from the stat call (linux, macos..) and on windows we would still maintain the millisecond precision as the _wstat64 call on Windows only supports that.

**Comment by sigurdm** [comment]
https://dart-review.googlesource.com/c/sdk/+/507341 added support down to microseconds on supported platforms. @a-siva  is right, that we cannot support nanoseconds without changing the DateTime class.

@lrhn is it possible/realistic to change DateTime to support also nanoseconds?

**Comment by lrhn** [comment]
`DateTime` with `nonosecondsSinceEpoch`?

TL;DR: Possible, but not in a _nice_ way.

The class is not `final`, so someone could be implementing it. (I think someone is, to add nanonseconds! Or was that `Duration`.)
So it's breaking to add a `nanosecondsSinceEpoch` getter.

If someone is implementing it, we can't really add an extension method either, that just won't work for them, and they'll have no way to integrate with it. Extension methods on our own classes is a design smell. It means we're stuck because we can't extend something, even if we think it would be better.

I think the "best" (read: safest and most likely to succeed) solution would be to introduce a new subclass, say (tongue-in-cheek) `DateTimeEx`, which implements `DateTime`, and which is the type actually returned by `DateTime` constructors.

You can use `DateTimeEx` directly, instead of `DateTime`, and you can safely cast all platform `DateTime` objects to `DateTimeEx`. You can't get to the nanoseconds unless you do an `myDate is DateTimeEx`, but you also don't need to know that it exists.

So if we _really_ want to, we probably can.

It won't work on JS, their `Date` doesn't have more than milliseconds, but that hasn't stopped us from having microseconds.
So `nanosecondsSinceEpoch` will just return `millisecondsSinceEpoch * 1_000_000`.

We also need nanoseconds on `Duration`, which also isn't `final`. So we'll need a `DurationEx` subtype with nanoseconds which `DateTimeEx.difference` returns.

We need to allow more digits after the `.` when parsing, which shouldn't be a problem.

For representation, the limits of DateTime is 100 million days on each side of the epoch.
100 million days is +/- 8_640_000_000_000_000_000 microseconds, aka 0x77e7_7239_2b60_0000. That fits very-very snugly in a 64-bit signed number. (JS is limited to +/- 53 bits, which is where 100 million days of milliseconds fit.)

So adding microseconds means adding a second integer to the representation.
At that point, we might as well split it into days-since-epoch + microseconds-in-day and make calendar-day calculations easier and less likely to overflow. (If we limit every argument of the `DateTime` constructor to +/- 100 million days, except nanoseconds, then cutting them into days and nanonseconds-in-day won't overflow the intermediate day sums!)

We'll need the same representation fix for `Duration`, which may then finally be able to represent the difference between min-day and max-day (which are 64-bits of microseconds apart).

All in all: Doable, but requires something we've tried to avoid so far, an "extended subtype" while retaining the original type, so there are now two ways to do the same thing. And if everybody moves to using `DateTimeEx`, and we can eventually remove `DateTime`, we _then_ have to make everybody rename back to the good name.
(So a new good name would be better. Not that `DateTime` is great. Is it a time-stamp, a calendar day or something third?)

EDIT: Caveat: The `DateTime` class is extensible and has public generative constructors.
That means we can't make those factory constructors that create `DateTimeEx` instances.
May need a breaking change anyway, and then we might as well make `DateTime` and `Duration` final.
